src/resqml2_2/AbstractColorMap.cpp

```diff
@@ -0,0 +1,152 @@
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+-----------------------------------------------------------------------*/
+#include "AbstractColorMap.h"
+
+#include "common/GraphicalInformationSet.h"
+
+using namespace std;
+using namespace gsoap_eml2_2;
+using namespace COMMON_NS;
+using namespace RESQML2_2_NS;
+
+void AbstractColorMap::setRgbColors(unsigned int colorCount,
+	double const* rgbColors, double const* alphas, vector<string> const& colorTitles,
+	double const* indices)
+{
+	double* hsvColors = new double[((double)colorCount) * 3];
+	for (size_t colorIndex = 0; colorIndex < colorCount; ++colorIndex) {
+		if (rgbColors[3 * colorIndex] < 0 || rgbColors[3 * colorIndex] > 1) {
+			throw invalid_argument("red must be in range [0, 1]");
+		}
+
+		if (rgbColors[3 * colorIndex + 1] < 0 || rgbColors[3 * colorIndex + 1] > 1) {
+			throw invalid_argument("green must be in range [0, 1]");
+		}
+
+		if (rgbColors[3 * colorIndex + 2] < 0 || rgbColors[3 * colorIndex + 2] > 1) {
+			throw invalid_argument("blue must be in range [0, 1]");
+		}
+
+		GraphicalInformationSet::rgbToHsv(rgbColors[3 * colorIndex], rgbColors[3 * colorIndex + 1], rgbColors[3 * colorIndex + 2],
+			hsvColors[3 * colorIndex], hsvColors[3 * colorIndex + 1], hsvColors[3 * colorIndex + 2]);
+	}
+
+	setHsvColors(colorCount, hsvColors, alphas, colorTitles, indices);
+
+	delete[] hsvColors;
+}
+
+void AbstractColorMap::setRgbColors(unsigned int colorCount,
+	unsigned int const* rgbColors, double const* alphas, vector<string> const& colorTitles,
+	double const* indices)
+{
+	double* hsvColors = new double[((double)colorCount) * 3];
+	for (size_t colorIndex = 0; colorIndex < colorCount; ++colorIndex) {
+		if (rgbColors[3 * colorIndex] < 0 || rgbColors[3 * colorIndex] > 255) {
+			throw invalid_argument("red must be in range [0, 255]");
+		}
+
+		if (rgbColors[3 * colorIndex + 1] < 0 || rgbColors[3 * colorIndex + 1] > 255) {
+			throw invalid_argument("green must be in range [0, 255]");
+		}
+
+		if (rgbColors[3 * colorIndex + 2] < 0 || rgbColors[3 * colorIndex + 2] > 255) {
+			throw invalid_argument("blue must be in range [0, 255]");
+		}
+
+		GraphicalInformationSet::rgbToHsv(rgbColors[3 * colorIndex], rgbColors[3 * colorIndex + 1], rgbColors[3 * colorIndex + 2],
+			hsvColors[3 * colorIndex], hsvColors[3 * colorIndex + 1], hsvColors[3 * colorIndex + 2]);
+	}
+
+	setHsvColors(colorCount, hsvColors, alphas, colorTitles, indices);
+
+	delete[] hsvColors;
+}
+
+double AbstractColorMap::getHue(double colorIndex) const
+{
+	resqml2__HsvColor const* const color = getColor(colorIndex);
+	if (color == nullptr) {
+		throw invalid_argument("There is no such color index");
```

Comment:
an out_of_range exception would be better imho.

Reply:
Done